src/RoslynPad.Common.UI/Services/DocumentFileWatcher.cs

```diff
@@ -0,0 +1,114 @@
+
+        private class Unsubscriber<T> : IDisposable
+        {
+            private readonly List<IObserver<T>> _observers;
+            private readonly IObserver<T> _observer;
+
+            public Unsubscriber(List<IObserver<T>> observers, IObserver<T> observer)
+            {
+                _observers = observers;
+                _observer = observer;
+            }
+            public void Dispose() => _observers.Remove(_observer);
+        }
+
+        [ImportingConstructor]
+        public DocumentFileWatcher(IAppDispatcher appDispatcher)
+        {
+            _appDispatcher = appDispatcher;
+            _fileSystemWatcher = new FileSystemWatcher();
+            _fileSystemWatcher.Created += OnChanged;
+            _fileSystemWatcher.Renamed += OnRenamed;
+            _fileSystemWatcher.Deleted += OnChanged;
+            _fileSystemWatcher.IncludeSubdirectories = true;
+        }
+
+        public string Path
+        {
+            get => _fileSystemWatcher.Path;
+            set
+            {
+                _fileSystemWatcher.Path = value;
+                _fileSystemWatcher.EnableRaisingEvents = IOUtilities.IsDirectory(_fileSystemWatcher.Path);
+                _observers.Clear(); //Most likely root has changed
+            }
+        }
+
+        private void OnChanged(object sender, FileSystemEventArgs e) => Publish(new DocumentFileChanged(ToDocumentFileChangeType(e.ChangeType), e.FullPath));
+
+        private DocumentFileChangeType ToDocumentFileChangeType(WatcherChangeTypes changeType)
+        {
+            switch (changeType)
+            {
+                case WatcherChangeTypes.Created:
+                    return DocumentFileChangeType.Created;
+                case WatcherChangeTypes.Deleted:
+                    return DocumentFileChangeType.Deleted;
+                case WatcherChangeTypes.Renamed:
+                    return DocumentFileChangeType.Renamed;
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(changeType), changeType, null);
+            }
+        }
+
+        private void OnRenamed(object sender, RenamedEventArgs e) => Publish(new DocumentFileChanged(ToDocumentFileChangeType(e.ChangeType), e.OldFullPath, e.FullPath));
+
+        private void Publish(DocumentFileChanged documentFileChanged)
+        {
+            foreach (var observer in _observers.ToArray())
+            {
+                _appDispatcher.InvokeAsync(() => observer.OnNext(documentFileChanged));
+            }
+        }
+
+        public void Dispose()
+        {
+            _fileSystemWatcher?.Dispose();
+            GC.SuppressFinalize(this);
+        }
+
+        public IDisposable Subscribe(IObserver<DocumentFileChanged> observer)
+        {
+            if (!_observers.Contains(observer))
+                _observers.Add(observer);
+
+            return new Unsubscriber<DocumentFileChanged>(_observers, observer);
+        }
+
+        ~DocumentFileWatcher() => Dispose();
```

Comment:
Remove this (and `GC.SuppressFinalize` above). A finalizer is not required if the disposed object is a managed object.